File: python/S14_BST/L2_PROBLEMS/9_inorderpredecessor.c
#include <stdio.h>
#include <stdlib.h>
struct node
{
    int data;
    struct node *left;
    struct node *right;
};
struct node *createnode(int data)
{
    struct node *newnode = (struct node *)malloc(sizeof(struct node));
    newnode->data = data;
    newnode->left = NULL;
    newnode->right = NULL;
    return newnode;
}
struct node *max(struct node *root)
{
    if (root == NULL)
        return NULL;
    while (root && root->right)
    {
        root = root->right;
    }
    return root;
}
struct node *insert(struct node *root,int data)
{
    if(root==NULL)
    {
        root=createnode(data);
        return root;
    }
    if(root->data>data)
    {
        root->left=insert(root->left,data);
    }
    else if(root->data<data)
    {
        root->right=insert(root->right,data);
    }
    return root;
}
struct node *inorderpredecessor(struct node *root, int data)
{
    if (root == NULL)
        return NULL;
    struct node *curr = root;
    struct node *pre = NULL;
    while (curr!=NULL)
    {
        if (curr->data > data)
        {
            
            curr = curr->left;
        }
        else if (curr->data < data)
        {
            pre=curr;
            curr = curr->right;
        }
        else
        {
            if (curr->left)
                return max(curr->left);
            else
                break;
        }
    }
    return pre;
}
int main()
{
    struct node *root=NULL;
    root=insert(root,67);
    root=insert(root,89);
    root=insert(root,809);
    root=insert(root,890);
    root=insert(root,829);
    struct node *pre=inorderpredecessor(root,89);
    if(pre==NULL)printf("not exist");
    printf("%d",pre->data);
    return 0;

    
}

File: JAVA/S14_BST/L2_PROBLEMS/_9_inorderpredecessor.java
class Node{
    int data;
    Node left;
    Node right;
    Node(int data){
        this.data=data;
        this.left=null;
        this.right=null;
    }
}
class Bst{
    Node insert(Node root,int data){
        Node newnode=new Node(data);
        if(root==null){
            return newnode;
        }
        if(root.data>data){
            root.left=insert(root.left,data);
        }
        if(root.data<data){
            root.right=insert(root.right,data);
        }
        return root;

    }
    public void inorder(Node root){
        if(root==null){
            return;
        }
        inorder(root.left);
        System.out.print(root.data+" ");
        inorder(root.right);
    }
    Node max(Node root){
        if(root==null)return null;
        Node curr=root;
        while(curr!=null && curr.right !=null){
            curr=curr.right;
        }
        return curr;
    }
    Node predecessor(Node root,int data){
        if(root==null)return null;
        Node curr=root;
        Node pre=null;
        while(curr!=null){
            if(curr.data<data){
                pre=curr;
                curr=curr.right;
            }
            else if(curr.data>data){
                curr=curr.left;
            }
            else{
                if(curr.left!=null){
                    return max(curr.left);
                }
                else break;
            }
            
        }
        return pre;
    }
   
    
}
public class _9_inorderpredecessor{
    public static void main(String[]args){
        Node root=null;
        Bst b=new Bst();
        root=b.insert(root,67);
        root=b.insert(root,6);
        root=b.insert(root,607);
        root=b.insert(root,60);
        root=b.insert(root,617);
        root=b.insert(root,63);
        Node s=b.predecessor(root,60);
        // b.inorder(root);
        System.out.print(s.data);
    }
}